### DISCUSSION TITLE
[DuckDB Wasm] How to use named prepared statements?

### DISCUSSION THREAD

**Discussion by irgijs** [duckdb/duckdb#11782] [Q&A]
Does the DuckDB WASM API support named prepared statements? For instance, I would expect this to work.

```typescript
const connection = database.connect();

const statement = await connection.prepare("SELECT $foo >= 0");
const result = await statement.query({ foo: 1 });
```

However, the abovementioned code results in this error.

> Error: Invalid column type encountered for argument 0

Could someone explain the correct way to use a named prepared statement?

**Comment by Tishj** [comment]
I think WASM does not support it yet
@carlopi

**Comment by carlopi** [comment]
Have a look at this test: 
```ts
TEST(WebDB, PrepareQuery) {
    auto db = make_shared<WebDB>(NATIVE);
    WebDB::Connection conn{*db};
    auto stmt = conn.CreatePreparedStatement("SELECT ? + 5");
    ASSERT_TRUE(stmt.ok()) << stmt.status().message();
    auto buffer = conn.RunPreparedStatement(*stmt, "[4]");
    ASSERT_TRUE(buffer.ok()) << buffer.status().message();
    auto success = conn.ClosePreparedStatement(*stmt);
    ASSERT_TRUE(success.ok()) << success.message();
}
```
that should offer some guidance on prepared statements. Code is here: https://github.com/duckdb/duckdb-wasm/blob/main/lib/test/webdb_test.cc#L58

**Reply by Tishj** [reply]
They're asking about [named prepared parameters](https://duckdb.org/docs/sql/query_syntax/prepared_statements#named-parameters-parameter) specifically, but this looks like a different API entirely?

**Reply by irgijs** [reply]
Hi Carlo,

The test case you mention is related to an unnamed prepared statement (which works fine on the browser indeed).

However, I need to execute a named prepared statement. It appears to be partially supported since the underlying engine raises an issue about a missing value for the "foo" parameter (in my example) if no parameters are supplied at all.

I think the issue is that the javascript `query` method collects all the parameters into a [single array](https://github.com/duckdb/duckdb-wasm/blob/58fcb9a46b73eac1abb9b0dee9d7c46d1a84f628/packages/duckdb-wasm/src/bindings/connection.ts#L156), which is then [JSON stringified](https://github.com/duckdb/duckdb-wasm/blob/58fcb9a46b73eac1abb9b0dee9d7c46d1a84f628/packages/duckdb-wasm/src/bindings/bindings_base.ts#L315C1-L316C1) and forwarded to the wasm package as 1 parameter. This would explain why DuckDB complained about invalid column types.

**Reply by carlopi** [reply]
Right, no there is currently no support for NAMED prepared statement, but you can use the simpler version where arguments are positionally substituted from the query statement for each `?`.

```
const connection = database.connect();

const statement = await connection.prepare("SELECT $ >= 0");
const result = await statement.query(1);
```
